providers/subclassing_an_existing_provider.md

```diff
@@ -0,0 +1,189 @@
+
+### Creating the new plugin
+
+First step is to create a new empty plugin:
+```bash
+$ bundle exec rails generate manageiq:provider ManageIQ::Providers::AwesomePrivateCloud --no-scaffolding --vcr
+** ManageIQ master, codename: Oparin
+      create  
+         run  git init /home/grare/adam/src/manageiq/manageiq/plugins/manageiq-providers-awesome_private_cloud from "."
+
+Initialized empty Git repository in /home/grare/adam/src/manageiq/manageiq/plugins/manageiq-providers-awesome_private_cloud/.git/
+      create  manageiq-providers-awesome_private_cloud.gemspec
+      create  .codeclimate.yml
+      create  .gitignore
+      create  .rspec
+      create  .rspec_ci
+      create  .rubocop.yml
+      create  .rubocop_cc.yml
+      create  .rubocop_local.yml
+      create  .whitesource
+      create  .yamllint
+      create  Gemfile
+      create  LICENSE.txt
+      create  Rakefile
+      create  README.md
+      create  bin/ci/after_script
+      create  bin/rails
+      create  bin/setup
+      create  bin/update
+      create  bundler.d
+      create  bundler.d/.keep
+      create  config/secrets.defaults.yml
+      create  config/settings.yml
+      create  lib/manageiq-providers-awesome_private_cloud.rb
+      create  lib/manageiq/providers/awesome_private_cloud/engine.rb
+      create  lib/manageiq/providers/awesome_private_cloud/version.rb
+      create  lib/tasks/README.md
+      create  lib/tasks_private/spec.rake
+      create  locale
+      create  locale/.keep
+      create  spec/factories
+      create  spec/support
+      create  spec/spec_helper.rb
+      insert  /home/grare/adam/src/manageiq/manageiq/Gemfile
+      create  spec/models/manageiq/providers/awesome_private_cloud
+        gsub  lib/tasks_private/spec.rake
+      insert  .yamllint
+      append  spec/spec_helper.rb
+
+```
+
+Follow the steps from [Creating a new provider](writing_a_new_provider.md) for setting up your new plugin for local development by adding to your `bundler.d/override.rb` and symlinking your `spec/manageiq` directory.
+
+You'll notice however that we chose "--no-scaffolding" option, this is going to allow us to define the classes that we need as subclasses of the OpenStack provider.
+
+### Creating the subclasses
+
+Now that we have an empty plugin we have to start adding the subclasses that we need for the provider to work.
+
+First let's create the initial directories that will hold our subclasses:
+```bash
+$ mkdir -p app/models/manageiq/providers/awesome_private_cloud/{cloud_manager,network_manager,storage_manager} app/models/manageiq/providers/awesome_private_cloud/inventory/{collector,parser,persister}
+```
+
+Now we can start defining our main manager class that will inherit from OpenStack:
+```ruby
+ManageIQ::Providers::Openstack::CloudManager.include(ActsAsStiLeafClass)
```

Comment:
```suggestion
ManageIQ::Providers::AwesomeCloud::CloudManager.include(ActsAsStiLeafClass)
```

Reply:
OpenStack was intentional, the idea here was we had an on-premise provider which implements an OpenStack compatible API.

`AwesomeCloud` was our example of a public-cloud

Reply:
Oh I see...I think what's confusing is it's named AwesomePrivateCloud, and name-wise I thought it was a subclass of AwesomeCloud because they are similarly named.

Reply:
Yeah in terms of back-story I was going for "If AwesomeCloud wanted to release an on-premise version"